### DISCUSSION TITLE
Calculate total with fieldArray

### DISCUSSION THREAD

**Discussion by kaisiant** [react-hook-form/react-hook-form#3179] [Q&A] [answered]
Hello bro, need some help from you.

I want to calculate the total amount by quantity * price.
I am using watch and setValue together, and infinite loop happened.
Is there any way to allow me set the value?

Here is the CSB: https://codesandbox.io/s/react-hook-form-usefieldarray-forked-6ohxx?file=/src/index.js

**Comment by bluebill1049** [comment]
you want to take a look at this example: https://codesandbox.io/s/watchusewatch-calc-tckxm maybe it will be helpful.

**Reply by kaisiant** [reply]
The example only displays the total, what if I want to set the total to the current row item.
For example, `item[i].total`

I have another question in my mind, we are not possible to set the value to the form if we are not using any controller or register, is it?
For example, I want to `setValue('test', testValue) `but the fieldName `'test'` is not registered, can I still able to see the value in the form when submitting?

**Reply by bluebill1049** [reply]
-  you can apply a similar idea for row
- question is the input get to register?

**Reply by kaisiant** [reply]
Okay I get your point.

Here is what I have done.

 ```
 const watchTest = useWatch({
    control,
    name: "test",
    defaultValue: fields
  });
```
Get the total
```
 const totalAmount =
            watchTest[index].quantity * watchTest[index].price;
```

But I'm sucked on next move(set the value)
```
  <Controller
                as={<Input />}
                name={`test[${index}].totalAmount`}
                control={control}
                defaultValue={item.totalAmount}
 />
```

**Reply by kaisiant** [reply]
Hi, is there any update for this.

**Reply by bluebill1049** [reply]
hey @kaisian97 you can use `setValue` API right?

**Reply by kaisiant** [reply]
> hey @kaisian97 you can use `setValue` API right?

I have no idea where I should place it, both inside the map or useEffect will causing infinite loop
![image](https://user-images.githubusercontent.com/26655111/96418763-d44f7880-1225-11eb-8cee-d537adbceb72.png)

**Reply by bluebill1049** [reply]
what about this? https://codesandbox.io/s/watchusewatch-calc-tckxm?file=/src/Calc.js

**Reply by kaisiant** [reply]
> https://codesandbox.io/s/watchusewatch-calc-tckxm?file=/src/Calc.js

It's different, the example is trying to get the grant total and not setting the value inside the field array so it won't have infinite loop. What I need is calculating each item in the array and set the value in the array as well. 

The question for me is how to set the value inside the field array when watching the array?

**Reply by bluebill1049** [reply]
What's the difference here? you can build a child component which watches the FieldArray and then just `setValue`?

**Reply by bluebill1049** [reply]
you can put the calculate component inside the FieldArray (in my example), it's still going to work right?

**Reply by bluebill1049** [reply]
let me try it.

**Reply by bluebill1049** [reply]
https://codesandbox.io/s/watchusewatch-calc-forked-sswmw?file=/src/Calc.js

**Reply by kaisiant** [reply]
> https://codesandbox.io/s/watchusewatch-calc-forked-sswmw?file=/src/Calc.js

I think you are misunderstood. You are still setting the value outside the fieldArray, that is not an issue. My use case is need to set the value in the array and the array is being watched.

For example: 
```
price | quantity | total
 1             5              5
 2             5             10

const total = watchTest.quantity * watchTest.price
setValue('test[i].total, total)
```
I want to set the total to that current row directly instead of displaying the total only.

**Reply by bluebill1049** [reply]
oh... i see let me think about it.

**Reply by bluebill1049** [reply]
- definitely not going to work during the render
- have you consider to setValue at `useEffect`? place those compare logic into a component and update input at useEffect inside

**Comment by bluebill1049** [comment, marked as answer]
`setValue(item.totalAmount, totalAmount);` which field you are try to set?

**Reply by kaisiant** [reply]
> `setValue(item.totalAmount, totalAmount);` which field you are try to set?

I want to set the total of price * quantity in the array[i].totalAmount

**Comment by kaisiant** [comment]
Okay, I found the solution. Sorry for bothering you haha

Should set the value on onChange 
```

 const setTotal = (i, quantity, price) => {
    const amount = quantity * price;
    setValue(`test[${i}].totalAmount`, amount);
  };

           <Controller
                control={control}
                name={`test[${index}].quantity`}
                defaultValue={`${item.quantity}`}
                render={({ value, onChange, ...restProps }) => (
                  <input
                    value={value}
                    onChange={(e) => {
                      const quantity = e.target.value;
                      setTotal(index, quantity, watchTest[index].price);
                      onChange(quantity);
                    }}
                    {...restProps}
                  />
                )}
              />
              <Controller
                render={({ value, onChange, ...restProps }) => (
                  <input
                    value={value}
                    onChange={(e) => {
                      const price = e.target.value;
                      setTotal(index, watchTest[index].quantity, price);
                      onChange(price);
                    }}
                    {...restProps}
                  />
                )}
                name={`test[${index}].price`}
                control={control}
                defaultValue={item.price}
              />
```

**Reply by bluebill1049** [reply]
aha! no worries at all! good solution.

**Reply by matt2574** [reply]
Can someone share a working example of this? I'm trying to accomplish the same thing with a nested array and not working.

**Reply by abesamislyndon** [reply]
> Can someone share a working example of this? I'm trying to accomplish the same thing with a nested array and not working.

hi have you solve this function? I have the same issue. hope can help me on this. thanks

**Comment by abesamislyndon** [comment]
Hi anyone for complete function on this? thanks

**Comment by neyric** [comment]
I don't think rendering the total by setValue is a good way to do it.
**Try this:** 
``` JSX
<Controller 
  control={control} 
  name={`test[${index}].quantity`}
  render={({ value: quantity }) => (
    <Controller 
      control={control} 
      name={`test[${index}].price`}
      render={({ value: price }) => {
        const total = quantity * price
        return <div>{total}</div>
      }}
    />
  )}
/>
```
This use two Controllers to get value for price and quantity and calc it in render.

**Comment by shkreios** [comment]
This is a working solution with `react-hook-form` `7.x.x` where you can use watch inside `useEffect` to subscribe imperativly
https://stackblitz.com/edit/github-znkgon?file=src/App.tsx
```tsx
const { control, register, watch, setValue } = useForm({
    defaultValues: {
      array: new Array<{ quantity: number; price: number; total: number }>(),
    },
  });

  const { fields, append } = useFieldArray({ control, name: 'array' });

  useEffect(() => {
    const { unsubscribe } = watch((value, info) => {
      if (
        info.name?.startsWith('array') &&
        (info.name?.endsWith('price') || info.name?.endsWith('quantity'))
      ) {
        const index = parseInt(info.name.split('.')[1]);
        const price = value.array?.at(index)?.price;
        const quantity = value.array?.at(index)?.quantity;
        if (price && quantity) setValue(`array.${index}.total`, price * quantity);
      }
    });
    return () => unsubscribe();
  }, [watch]);

```

**Reply by shkreios** [reply]
Personally, I think if you will not attach the total amount to its input field it, then it doesn't make sense to track the value in `rhf` and calculating it inline is fine. 

But if, for example, you have the requirement that that all 3 values should be editable and editing one should reactively update the other 2 this solution is nice.

**Reply by shkreios** [reply]
This is example on how implementation would look for this use case
https://stackblitz.com/edit/github-znkgon-tstd5g?file=src/App.tsx

```tsx
useEffect(() => {
    const { unsubscribe } = watch((_, { name, type }) => {
      // first get all values
      const value = getValues();
      // check if the update was a change not a delete
      if (type === 'change' && name) {
        // check if a subproperty has changed and not the array or some of its elements themself
        if (endsWith(name, 'quantity') || endsWith(name, 'total') || endsWith(name, 'price')) {
          const { array } = value;
          // get the index and the name of field that has been changed
          const [, indexString, fieldName] = name.split('.');

          // Get the new value from the field
          const fieldValue = get(value, name) as FieldPathValue<typeof value, typeof name>;

          const index = parseInt(indexString);

          if (fieldValue) {
            if (fieldName === 'quantity')
              setValue(`array.${index}.total`, fieldValue * array[index].price);
            else if (fieldName === 'price')
              setValue(`array.${index}.total`, fieldValue * array[index].quantity);
            else if (fieldName === 'total')
              setValue(`array.${index}.price`, fieldValue / array[index].quantity);
          }
        }
      }
    });
    return () => unsubscribe();
  }, [watch]);
```

**Comment by adamwozhere** [comment]
I might be perhaps thinking about this the wrong way, I'm quite new to react-hook-form, but I'm using Zod for my schema and have an array of values, with a `.transform()` in the schema to calculate the item total:
```
items: z
    .array(
      z
        .object({
          name: z.string().nonempty(),
          quantity: z.coerce.number().min(1),
          price: z.coerce.number(),
          total: z.coerce.number(),
        })
        .transform((val) => {
          return {
            ...val,
            total: val.price * val.quantity,
          };
        })
...
```
Is there perhaps a way to retrieve all the values from the zodResolver rather than just the boolean success values? That way the Total value is enforced/created in the schema itself and only calculated once, rather than calculated in UI and validated in the schema.
